Categorized exclusions

spriadka opened the pull request:
<!-- 
Many thanks for contributing to Arquillian! Together we can make the testing world better.

Please tell us what this PR brings following the template we provided. 
And don't forget to link to the issue (or create one if there is none).

If you are still working on the change please prefix this pull request title with "WIP"

YOU CAN DELETE THIS COMMENT :)
-->

#### Short description of what this resolves:

This PR implements the possibility to manually exclude test classes/methods from execution based on their category (see #310 )

#### Changes proposed in this pull request:

- added the ``excludedCategories`` into the ``categorized`` configuration
- changed the ``AbstractParser`` logic to support this feature
- added the ``CategorizedTestSelection`` which serves as an information holder to the categorized selection, which then transforms to the ``TestSelection`` using the ``toTestSelection()`` method.


Fixes #

spriadka commented:
@MatousJobanek WDYT?

MatousJobanek commented:
Hi @spriadka 
thanks for your contribution and sorry for my long inactivity - I was knocked out by flu in last 8 days. I'll do the review today.

MatousJobanek commented on strategies/categorized/src/main/java/org/arquillian/smart/testing/strategies/categorized/CategorizedTestSelection.java:
```diff
@@ -0,0 +1,57 @@
+            return new TestSelection(className, appliedStrategies);
+        }
+        List<String> filteredTestMethodNames = new ArrayList<>(includedTestMethodNames);
+        filteredTestMethodNames.removeAll(excludedTestMethodNames);
```
if you exclude all test methods, then it should return `TestSelection.NOT_MATCHED`, but you return `TestSelection` with an empty list of the test methods which means that it should execute the whole test class.

spriadka replied:
Thanks for the feedback! Removed whole `CategorizedTestSelection` class in d1806ed

MatousJobanek commented on strategies/categorized/src/test/java/org/arquillian/smart/testing/strategies/categorized/CategorizedTestsDetectorTest.java:
```diff
@@ -213,6 +208,94 @@ public void should_return_classes_and_methods_without_first_category_when_revers
+    public void should_not_exclude_test_class_when_excluding_single_category_of_many() {
+        //given
+        categorizedConfig.setCategories(new String[] {FirstCategory.class.getSimpleName()});
+        categorizedConfig.setExcludedCategories(new String[] {SecondCategory.class.getSimpleName()});
```
The test is called: `should_not_exclude_test_class_when_excluding_single_category_of_many` and you exclude the `SecondCategory` which is contained in `FirstAndSecondCategorizedClass`. I thought that if the class/method has the category that is excluded, then it shouldn't be executed at all. it is also described here: http://maven.apache.org/surefire/maven-surefire-plugin/test-mojo.html#excludedGroups

MatousJobanek commented on strategies/categorized/src/test/java/org/arquillian/smart/testing/strategies/categorized/CategorizedTestsDetectorTest.java:
```diff
@@ -213,6 +208,94 @@ public void should_return_classes_and_methods_without_first_category_when_revers
+        Collection<TestSelection> testSelection =
+            new CategorizedTestsDetector(config).selectTestsFromClasses(Collections.singleton(FirstAndSecondCategorizedClass.class));
+
+        assertThat(testSelection).containsTestClassSelectionsExactlyInAnyOrder(getTestSelection(FirstAndSecondCategorizedClass.class));
```
`should_exclude_test_class_when_excluding_multiple_categories_and_match_all_is_set` if the class should be excluded then the resulting test selection collection should contain anything - should be empty.

MatousJobanek commented on strategies/categorized/src/test/java/org/arquillian/smart/testing/strategies/categorized/CategorizedTestsDetectorTest.java:
```diff
@@ -213,6 +208,94 @@ public void should_return_classes_and_methods_without_first_category_when_revers
+    }
+
+    @Test
+    public void should_exclude_methods_when_matched_and_class_is_not_annotated() {
```
this test method doesn't test the exclude logic - the resulting set of methods will contain always only the method `testWithFirstCategory` because only this one has the `FirstCategory` specified in included categories.

spriadka commented on strategies/categorized/src/test/java/org/arquillian/smart/testing/strategies/categorized/CategoriesParserTest.java:
```diff
@@ -147,6 +147,7 @@ public void should_return_false_for_class_not_containing_all_of_the_set_categori
     private CategorizedConfiguration prepareConfig(String... categories) {
         CategorizedConfiguration categorizedConfig = new CategorizedConfiguration();
         categorizedConfig.setCategories(categories);
+        categorizedConfig.setExcludedCategories(new String[0]);
```
@MatousJobanek I was getting NPE's on all tests from `CategoriesParserTest` due to the `excludedCategories` initialization. Calling this explicitly solved that issue. However in the `CategorizedTestsDetector` test case, the `@Before` block initializes the configuration correctly and the `excludedCategories` are empty.

spriadka commented on strategies/categorized/src/main/java/org/arquillian/smart/testing/strategies/categorized/AbstractParser.java:
```diff
@@ -2,9 +2,10 @@
 import java.lang.reflect.Method;
-import java.util.Arrays;
-import java.util.List;
+import java.util.*;
```
Thanks for the feedback. Unfortunately this was a mistake generated by the IDE and it slipped out of my sight unnoticed. Fixed in d25fb23

spriadka commented on strategies/categorized/src/main/java/org/arquillian/smart/testing/strategies/categorized/AbstractParser.java:
```diff
@@ -13,75 +14,180 @@
+        parseExcludedMethodNames(categorizedTestSelection, categorizedMethods);
+        setInheritedFlagIfNeeded(clazz, categorizedTestSelection);
+        if (!categorizedTestSelection.getIncludedTestMethodNames().isEmpty()) {
+            return categorizedTestSelection.toTestSelection();
```
Thanks for the comment, your proposed method is more effective and I got rid of the `CategorizedTestSelection` class as well. See d1806ed

spriadka commented on strategies/categorized/src/test/java/org/arquillian/smart/testing/strategies/categorized/CategorizedTestsDetectorTest.java:
```diff
@@ -2,17 +2,12 @@
-import org.arquillian.smart.testing.strategies.categorized.project.categories.SecondCategory;
-import org.arquillian.smart.testing.strategies.categorized.project.categories.ThirdCategorizedClass;
-import org.arquillian.smart.testing.strategies.categorized.project.categories.WithCategorizedMethodsClass;
+import org.arquillian.smart.testing.strategies.categorized.project.categories.*;
```
Thanks fixed in d25fb23

spriadka commented:
Thanks for the feedback @MatousJobanek . Sorry for my long inactivity as well, I was not able to make until yesterday. I made some major changes e.g removed group imports, changed test logic, removed the `CategorizedTestSelection` class etc. However, I would like to stand by the `IncludingCategoryFilter` and `ExcludingCategoryFilter` as it seems more elegant way than passing a boolean parameter into a method.

MatousJobanek commented on strategies/categorized/src/main/java/org/arquillian/smart/testing/strategies/categorized/AbstractParser.java:
```diff
@@ -13,75 +19,154 @@
+            .map(method -> new CategorizedMethod(method.getName(), findCategories(method.getAnnotations())))
+            .collect(Collectors.toList());
+        inheritClassLevelCategoriesIfAny(clazz, categorizedMethods);
+        List<String> applicableMethodNames = categorizedMethods.stream()
```
here you iterate over the same list of `CategorizedMethod`s twice - in two streams, between these two iterations you feed the `CategorizedMethod`s with the class annotations. 
See my proposal https://github.com/arquillian/smart-testing/pull/311#discussion_r164986429 the idea was that adding the class categories should be part of the `CategorizedMethod` constructor. Another thing to notice is that you retrieve the class categories for every iteration separately https://github.com/arquillian/smart-testing/pull/311/files#diff-b179c0d6894a22602721a54a8efc1d54R70 but this is not necessary as the list of specified categories is always the same. in my proposal is retrieving the class categories once before the loop and feed the  `CategorizedMethod` with it

MatousJobanek commented on strategies/categorized/src/test/java/org/arquillian/smart/testing/strategies/categorized/CategoriesParserTest.java:
```diff
@@ -76,7 +76,7 @@ public void should_return_true_for_class_containing_all_of_the_set_categories_wh
 
         // when
-        TestSelection selection = categoriesParser.getTestSelectionIfMatched(FirstAndSecondCategorizedClass.class);
+        TestSelection selection =            categoriesParser.getTestSelectionIfMatched(FirstAndSecondCategorizedClass.class);
```
please delete the empty spaces
